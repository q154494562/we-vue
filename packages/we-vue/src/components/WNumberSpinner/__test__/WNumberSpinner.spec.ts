import { shallowMount } from '@vue/test-utils'
import NumberSpinner from '../'

describe('number-spinner', () => {
  test('create', () => {
    const wrapper = shallowMount(NumberSpinner, {
      propsData: {},
    })

    expect(wrapper.name()).toBe('w-number-spinner')
    expect(wrapper.classes()).toContain('wv-number-spinner')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('blur event', () => {
    const wrapper = shallowMount(NumberSpinner, {
      data: function () {
        return {
          currentValue: '',
        }
      },
    })

    wrapper.find('input').trigger('blur')

    expect(wrapper.vm.currentValue).toBe(wrapper.vm.min)
  })

  test('onChange method', () => {
    const wrapper = shallowMount(NumberSpinner, {
      propsData: {
        value: 1,
      },
    })

    wrapper.vm.$refs.input.value = '2'

    wrapper.find('input').trigger('change')

    expect(wrapper.vm.currentValue).toBe(2)
  })

  test('onPaste method', () => {
    const wrapper = shallowMount(NumberSpinner, {
      propsData: {
        value: 1,
        fillable: false,
      },
    })

    const mockEvent = new Event('paste')
    jest.spyOn(mockEvent, 'preventDefault')

    wrapper.vm.onPaste(mockEvent as ClipboardEvent)

    expect(mockEvent.preventDefault).toHaveBeenCalled()
  })

  test('onKeypress method when fillable is false', () => {
    const spy = jest.fn()

    const wrapper = shallowMount(NumberSpinner, {
      propsData: {
        value: 1,
        fillable: false,
      },
    })

    const mockEvent = {
      preventDefault: spy,
    }

    wrapper.find('input').trigger('keypress', mockEvent)

    expect(spy).toHaveBeenCalled()
  })

  test('click minus button', () => {
    let wrapper = shallowMount(NumberSpinner, {
      propsData: {
        value: 5,
      },
    })

    wrapper.find('.btn-minus').trigger('click')
    expect(wrapper.vm.currentValue).toBe(5 - wrapper.vm.step)

    // when the initial value is NaN
    wrapper = shallowMount(NumberSpinner, {
      propsData: {
        value: NaN,
      },
    })

    wrapper.find('.btn-minus').trigger('click')
    expect(wrapper.vm.currentValue).toBe(0)
  })

  test('click plus button', () => {
    let wrapper = shallowMount(NumberSpinner, {
      propsData: {
        value: 1,
      },
    })

    wrapper.find('.btn-plus').trigger('click')
    expect(wrapper.vm.currentValue).toBe(1 + wrapper.vm.step)

    // when the initial value is NaN
    wrapper = shallowMount(NumberSpinner, {
      propsData: {
        value: NaN,
      },
    })

    wrapper.find('.btn-plus').trigger('click')
    expect(wrapper.vm.currentValue).toBe(1)
  })

  test('watch currentValue', () => {
    const wrapper = shallowMount(NumberSpinner, {
      propsData: {},
    })

    wrapper.setData({
      currentValue: 'current-value',
    })

    expect(wrapper.emitted().input).toBeTruthy()
    expect(wrapper.emitted().change).toBeTruthy()
  })

  test('watch value', () => {
    const wrapper = shallowMount(NumberSpinner, {
      propsData: {
        min: 1,
        max: 10,
      },
    })

    wrapper.setProps({
      value: 2,
    })

    expect(wrapper.vm.currentValue).toBe(2)

    // set a value bigger than max
    wrapper.setProps({
      value: 11,
    })

    expect(wrapper.vm.currentValue).toBe(10)

    // set a value smaller than min
    wrapper.setProps({
      value: 0,
    })

    expect(wrapper.vm.currentValue).toBe(1)

    // set an empty string
    wrapper.setProps({
      value: '',
    })

    expect(wrapper.vm.currentValue).toBe(1)
  })
})
