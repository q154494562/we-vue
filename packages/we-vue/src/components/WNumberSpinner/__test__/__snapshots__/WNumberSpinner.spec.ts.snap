// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number-spinner create 1`] = `<div class="wv-number-spinner"><button disabled="disabled" class="spinner-btn btn-minus"></button><input type="number" min="0" max="100" step="1" style="width: 3em; text-align: center;"><button class="spinner-btn btn-plus"></button></div>`;
